crates/api_common/src/request.rs

```diff
@@ -63,47 +64,40 @@ pub async fn fetch_link_metadata(url: &Url, context: &LemmyContext) -> LemmyResu
     .await?
     .error_for_status()?;
 
-  // In some cases servers send a wrong mime type for images, which prevents thumbnail
-  // generation. To avoid this we also try to guess the mime type from file extension.
-  let content_type = mime_guess::from_path(url.path())
-    .first()
-    // If you can guess that its an image type, then return that first.
-    .filter(|guess| guess.type_() == mime::IMAGE)
-    // Otherwise, get the content type from the headers
-    .or(
-      response
-        .headers()
-        .get(CONTENT_TYPE)
-        .and_then(|h| h.to_str().ok())
-        .and_then(|h| h.parse().ok()),
-    );
-
-  let opengraph_data = {
-    // if the content type is not text/html, we don't need to parse it
-    let is_html = content_type
-      .as_ref()
-      .map(|c| {
-        (c.type_() == mime::TEXT && c.subtype() == mime::HTML)
-      ||
-      // application/xhtml+xml is a subset of HTML
-      (c.type_() == mime::APPLICATION && c.subtype() == "xhtml")
-      })
-      .unwrap_or(false);
-    if !is_html {
-      Default::default()
-    } else {
+  let mut content_type: Option<Mime> = response
+    .headers()
+    .get(CONTENT_TYPE)
+    .and_then(|h| h.to_str().ok())
+    .and_then(|h| h.parse().ok());
+
+  let mut opengraph_data = Default::default();
+
+  if let Some(c) = &content_type {
+    if (c.type_() == mime::TEXT && c.subtype() == mime::HTML)
+	    ||
+	    // application/xhtml+xml is a subset of HTML
+	    (c.type_() == mime::APPLICATION && c.subtype() == "xhtml")
+    {
       // Can't use .text() here, because it only checks the content header, not the actual bytes
       // https://github.com/LemmyNet/lemmy/issues/1964
       // So we want to do deep inspection of the actually returned bytes but need to be careful not
       // spend too much time parsing binary data as HTML
 
       // only take first bytes regardless of how many bytes the server returns
       let html_bytes = collect_bytes_until_limit(response, bytes_to_fetch).await?;
-      extract_opengraph_data(&html_bytes, url)
+      opengraph_data = extract_opengraph_data(&html_bytes, url)
         .map_err(|e| info!("{e}"))
-        .unwrap_or_default()
+        .unwrap_or_default();
     }
-  };
+    // If a server is serving `application/octet-stream`, it's likely a mistake,
+    // so we try to guess the file type from its magic number.
+    else if c.subtype() == "octet-stream" {
+      // Don't need to fetch as much data for this as we do with opengraph
+      let octet_bytes = collect_bytes_until_limit(response, 512).await?;
+      content_type = infer::get(&octet_bytes).map_or(content_type, |t| t.mime_type().parse().ok());
+    }
+  }
```

Comment:
Here is still the problem if `content-type` is None (eg if the server is down). So it would make sense to add an `else` with mime_guess.